modules/collect/src/main/java/com/opengamma/strata/collect/io/XmlFile.java

```diff
@@ -0,0 +1,282 @@
+      XmlElement parsed = children.isEmpty() ?
+          XmlElement.ofContent(elementName, attrs, content) :
+          XmlElement.ofChildren(elementName, attrs, children);
+      String ref = attrs.get(refAttr);
+      if (ref != null) {
+        refs.put(ref, parsed);
+      }
+      return parsed;
+
+    } catch (XMLStreamException ex) {
+      throw new IllegalArgumentException(ex);
+    }
+  }
+
+  // find the start element and parses the name
+  private static String parseElementName(XMLStreamReader reader) throws XMLStreamException {
+    int event = reader.getEventType();
+    while (event != XMLStreamConstants.START_ELEMENT) {
+      event = reader.next();
+    }
+    String prefix = reader.getPrefix();
+    if (prefix.length() == 0) {
+      return reader.getLocalName();
+    }
+    return prefix + ':' + reader.getLocalName();
+  }
+
+  // parses attributes into a map
+  private static ImmutableMap<String, String> parseAttributes(XMLStreamReader reader) {
+    ImmutableMap<String, String> attrs;
+    int attributeCount = reader.getAttributeCount() + reader.getNamespaceCount();
+    if (attributeCount == 0) {
+      attrs = ImmutableMap.of();
+    } else {
+      ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
+      for (int i = 0; i < reader.getAttributeCount(); i++) {
+        String prefix = reader.getAttributePrefix(i);
+        if (prefix.length() == 0) {
+          builder.put(reader.getAttributeLocalName(i), reader.getAttributeValue(i));
+        } else {
+          builder.put(prefix + ':' + reader.getAttributeLocalName(i), reader.getAttributeValue(i));
+        }
+      }
+      for (int i = 0; i < reader.getNamespaceCount(); i++) {
+        String namespacePrefix = reader.getNamespacePrefix(i);
+        if (namespacePrefix != null) {
+          builder.put("xmlns:" + namespacePrefix, reader.getNamespaceURI(i));
+        } else {
+          builder.put("xmlns", reader.getNamespaceURI(i));
+        }
+      }
+      attrs = builder.build();
+    }
+    return attrs;
+  }
+
+  //-------------------------------------------------------------------------
+  // creates the XML input factory., recreated each time to avoid JDK-8028111.
+  private static XMLInputFactory xmlInputFactory() {
+    XMLInputFactory factory = XMLInputFactory.newFactory();
+    factory.setProperty(XMLInputFactory.IS_COALESCING, true);
+    factory.setProperty(XMLInputFactory.IS_REPLACING_ENTITY_REFERENCES, true);
+    factory.setProperty(XMLInputFactory.IS_SUPPORTING_EXTERNAL_ENTITIES, false);
+    factory.setProperty(XMLInputFactory.SUPPORT_DTD, false);
+    return factory;
+  }
+
+  //-------------------------------------------------------------------------
+  /**
+   * Restricted constructor.
+   */
+  private XmlFile(XmlElement root, Map<String, XmlElement> refs) {
+    this.root = ArgChecker.notNull(root, "root");
+    this.refs = ImmutableMap.copyOf(ArgChecker.notNull(refs, "refs"));
```

Comment:
The call to `notNull` is redundant, `ImmutableMap.copyOf` will blow up if `refs` is null.